<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Standardized Test Analysis

# Vivek Bavda:
## DSIR 7/12
### Project 1 July 26, 2021:
### Analysis on moving the needle on high school outcomes by building political capital.












### Introduction:
*As the Mayor and Mayors around this state are aware, standardized tests provide only a snapshot of student performance. These snapshots often define whether the educational agenda of each of your programs is successful. With this in mind, this data analysis seeks to benchmark which test this city and the state of California should institute as a requirement for its college-bound students. The SAT and ACT  are compared to determine which test best reflects the achievement of California's students and which test results should be promoted to support your political agenda.*

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

The SAT and ACT are standardized tests that many colleges and universities in the United States require for their admissions process. This score is used along with other materials such as grade point average (GPA) and essay responses to determine whether or not a potential student will be accepted to the university. However, despite growth in locally determined outcomes, the SAT and ACT snapshots often are seen as the main evaluation point by your constituents. To the extent our students score well, it allows us greater political capital to institue educational reforms. It is imperative that we choose and mandate by state and local regulations which test demonstrates higher scores and bests represents our students' achievements.

This requires us to analyze the format and results of the SAT and ACT. The SAT has two sections of the test: Evidence-Based Reading and Writing and Math ([*source*](https://www.princetonreview.com/college/sat-sections)). The ACT has 4 sections: English, Mathematics, Reading, and Science, with an additional optional writing section ([*source*](https://www.act.org/content/act/en/products-and-services/the-act/scores/understanding-your-scores.html)).  The score rangesEDITThey have different score ranges, which you can read more about on their websites or additional outside sources (a quick Google search will help you understand the scores for each test):
* [SAT](https://collegereadiness.collegeboard.org/sat)
* [ACT](https://www.act.org/content/act/en.html)

It is important to also keep in mind that standardized tests have long been a controversial topic for students, administrators, and legislators. Since the 1940's, an increasing number of colleges have been using scores from students' performances on tests like the SAT and the ACT as a measure for college readiness and aptitude ([*source*](https://www.minotdailynews.com/news/local-news/2017/04/a-brief-history-of-the-sat-and-act/)). Supporters of these tests argue that these scores can be used as an objective measure to determine college admittance. Opponents of these tests claim that these tests are not accurate measures of students potential or ability and serve as an inequitable barrier to entry. Lately, more and more schools are opting to drop the SAT/ACT requirement for their Fall 2021 applications ([*read more about this here*](https://www.cnn.com/2020/04/14/us/coronavirus-colleges-sat-act-test-trnd/index.html)).
Even with this trend, our political calculus has not changed. Most people in our districts use SAT and ACT has shorthand. Therefore, we recomment you choose ACTSAT

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

### California ACT and SAT datasets from 2018 - 2019 were used in this analysis.

* [`act_2019_ca.csv`](./data/act_2019_ca.csv): 2019 ACT Scores in California by School, District, and County. 

    The total number of testtakers is listed.The average scores for the four parts of the test (English, Reading, Math, and Science) is listed. The composite score of the four parts is benchmarked by listing the number of students greater than or equal to the score 21 is also listed.

* [`sat_2019_ca.csv`](./data/sat_2019_ca.csv): 2019 SAT Scores in California by School, District and County. 

    The total number of testakers is listed. The dataset also breaks the scores down by grade level. Benchmarks are listed for the two measurement categories. First, the number of students who met or exceeded the benchmark for reading & writing is listed. Second, the number of students who met or exceeded the benchmark for math. There is also a category for meeting both benchmarks.


**To-Do:** *Fill out this cell with the datasets you will use for your analysis. Write a brief description of the contents for each dataset that you choose.*

### Outside Research

IF NO OUTSIDE RESEARCH< DELETEBased on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [4]:
# Mean Code:
meannumberstry=[5,9,3]
def mean_wo_np(meannumbers):
    return (sum(meannumbers))/len(meannumbers)
mean_wo_np(meannumberstry)

5.666666666666667

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [5]:
# Code: Standard Deviation Calculation
meannumberstry=[1,8,-4,9,6]
def standard_deviation(stdmeannumbers):
    return (sum([((stdmeannumber-mean_wo_np(stdmeannumbers))**2)\
            for stdmeannumber in stdmeannumbers])/len(stdmeannumbers)) ** .5


standard_deviation(meannumberstry)



4.857983120596447

3. Data cleaning function:
    
    Write a function that takes in a string that is a number and a percent symbol (ex. '50%', '30.5%', etc.) and converts this to a float that is the decimal approximation of the percent. For example, inputting '50%' in your function should return 0.5, '30.5%' should return 0.305, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [6]:
# Code for turning percents or whole numbers into decimals

def percent_decimal(percent_input):
    if '%' in percent_input:
        return (float(percent_input.strip('%')))/100
    else:
        return (float(percent_input))/100
percent_decimal('50.1%')


    
#Source used for this function:https://stackoverflow.com/questions/12432663/what-is-a-clean-way-to-convert-a-string-percent-to-a-float

0.501

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [7]:
# Imports:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#Data import of California 2018-2019 ACT and SAT datasets used in analysis
ca_sat_2019 = '../data/sat_2019_ca.csv'
ca_act_2019 = '../data/act_2019_ca.csv'





In [8]:
dfa = pd.read_csv(ca_act_2019)
dfs = pd.read_csv(ca_sat_2019)


## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind the minimum & maximum possible values for each test/subtest).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of percents (ex. '50%', '30.5%', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to floats! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [9]:
# Code for Data Import and Cleaning:

#First 5 rows of SAT data frame code is below
dfs.head()

CDS  CCode     CDCode      SCode RType                    SName  \
0  6.615981e+12    6.0   661598.0   630046.0     S  Colusa Alternative Home   
1  6.616061e+12    6.0   661606.0   634758.0     S          Maxwell Sr High   
2  1.964733e+13   19.0  1964733.0  1930924.0     S      Belmont Senior High   
3  1.964733e+13   19.0  1964733.0  1931476.0     S  Canoga Park Senior High   
4  1.964733e+13   19.0  1964733.0  1931856.0     S     Whitman Continuation   

                 DName        CName  Enroll12  NumTSTTakr12  ...  \
0       Colusa Unified       Colusa      18.0           0.0  ...   
1      Maxwell Unified       Colusa      29.0          10.0  ...   
2  Los Angeles Unified  Los Angeles     206.0         102.0  ...   
3  Los Angeles Unified  Los Angeles     227.0         113.0  ...   
4  Los Angeles Unified  Los Angeles      18.0          14.0  ...   

  NumERWBenchmark11 PctERWBenchmark11 NumMathBenchmark11 PctMathBenchmark11  \
0               NaN               NaN                NaN                NaN   
1                 *                 *                  *                  *   
2                42             24.14                 12               6.90   
3                97             35.27                 37              13.45   
4                 *                 *                  *                  *   

   TotNumBothBenchmark12  PctBothBenchmark12 TotNumBothBenchmark11  \
0                    NaN                 NaN                   NaN   
1                      *                   *                     *   
2                     14               13.73                    11   
3                     18               15.93                    35   
4                      *                   *                     *   

  PctBothBenchmark11     Year Unnamed: 25  
0                NaN  2018-19         NaN  
1                  *  2018-19         NaN  
2               6.32  2018-19         NaN  
3              12.73  2018-19         NaN  
4                  *  2018-19         NaN  

[5 rows x 26 columns]

In [10]:
#First 5 rows of ACT data frame code is below
dfa.head()

CDS  CCode     CDCode      SCode RType  \
0  3.366993e+13   33.0  3366993.0   129882.0     S   
1  1.964212e+13   19.0  1964212.0  1995596.0     S   
2  1.563776e+13   15.0  1563776.0  1530377.0     S   
3  4.369666e+13   43.0  4369666.0  4333795.0     S   
4  1.964733e+13   19.0  1964733.0  1935121.0     S   

                             SName                  DName        CName  \
0  21st Century Learning Institute       Beaumont Unified    Riverside   
1      ABC Secondary (Alternative)            ABC Unified  Los Angeles   
2      Abraham Lincoln Alternative  Southern Kern Unified         Kern   
3             Abraham Lincoln High       San Jose Unified  Santa Clara   
4      Abraham Lincoln Senior High    Los Angeles Unified  Los Angeles   

   Enroll12  NumTstTakr AvgScrRead AvgScrEng AvgScrMath AvgScrSci NumGE21  \
0      18.0         0.0        NaN       NaN        NaN       NaN     NaN   
1      58.0         0.0        NaN       NaN        NaN       NaN     NaN   
2      18.0         0.0        NaN       NaN        NaN       NaN     NaN   
3     463.0        53.0         23        22         22        23      34   
4     226.0        19.0         21        20         23        22      11   

  PctGE21     Year  Unnamed: 17  
0     NaN  2018-19          NaN  
1     NaN  2018-19          NaN  
2     NaN  2018-19          NaN  
3   64.15  2018-19          NaN  
4   57.89  2018-19          NaN

In [11]:
#checking data types
dfa.dtypes

CDS            float64
CCode          float64
CDCode         float64
SCode          float64
RType           object
SName           object
DName           object
CName           object
Enroll12       float64
NumTstTakr     float64
AvgScrRead      object
AvgScrEng       object
AvgScrMath      object
AvgScrSci       object
NumGE21         object
PctGE21         object
Year            object
Unnamed: 17    float64
dtype: object

In [12]:
#checking data types
dfs.dtypes

CDS                      float64
CCode                    float64
CDCode                   float64
SCode                    float64
RType                     object
SName                     object
DName                     object
CName                     object
Enroll12                 float64
NumTSTTakr12             float64
NumERWBenchmark12         object
PctERWBenchmark12         object
NumMathBenchmark12        object
PctMathBenchmark12        object
Enroll11                 float64
NumTSTTakr11             float64
NumERWBenchmark11         object
PctERWBenchmark11         object
NumMathBenchmark11        object
PctMathBenchmark11        object
TotNumBothBenchmark12     object
PctBothBenchmark12        object
TotNumBothBenchmark11     object
PctBothBenchmark11        object
Year                      object
Unnamed: 25              float64
dtype: object

In [13]:
dfs.isnull().sum()

CDS                         1
CCode                       1
CDCode                      1
SCode                       1
RType                       1
SName                     598
DName                      59
CName                       1
Enroll12                    1
NumTSTTakr12                1
NumERWBenchmark12         276
PctERWBenchmark12         276
NumMathBenchmark12        276
PctMathBenchmark12        276
Enroll11                    1
NumTSTTakr11                1
NumERWBenchmark11         311
PctERWBenchmark11         311
NumMathBenchmark11        311
PctMathBenchmark11        311
TotNumBothBenchmark12     276
PctBothBenchmark12        276
TotNumBothBenchmark11     311
PctBothBenchmark11        311
Year                        1
Unnamed: 25              2580
dtype: int64

In [14]:
dfa.isnull().sum()


CDS               1
CCode             1
CDCode            1
SCode           523
RType             1
SName           581
DName            59
CName             1
Enroll12          1
NumTstTakr        1
AvgScrRead      357
AvgScrEng       357
AvgScrMath      357
AvgScrSci       357
NumGE21         357
PctGE21         357
Year              1
Unnamed: 17    2310
dtype: int64

In [15]:
# Change the column names using the `.rename()` function for SAT
# Original column: renamed column
srename_map = {'CDS': 'county_district_school_code','CCode': 'county_code', 'CDCode': 'district_code','SCode':'school_code','RType': 'record_type'\
    , 'SName': 'school_name','DName': 'district_name', 'CName': 'county_name', 'Enroll12': '_12th_grade_enrollment', 'NumTSTTakr12': 'test_taker_count_12thgrade'\
               ,'NumERWBenchmark12': 'number_that_passed_reading_writing_standard_12thgrade', 'PctERWBenchmark12': 'percent_that_passed_reading_writing_standard_12thgrade', 'NumMathBenchmark12': 'number_that_passedorabove_math_standard_12thgrade'\
   , 'PctMathBenchmark12':'percent_that_passedorabove_math_standard_12thgrade', 'Enroll11':'_11th_grade_enrollment', 'NumTSTTakr11':'test_taker_count_11thgrade', 'NumERWBenchmark11': 'number_that_passed_reading_writing_standard_11thgrade','PctERWBenchmark11':'percent_that_passed_reading_writing_standard_11thgrade', 'NumMathBenchmark11':'number_that_passedorabove_math_standard_11thgrade'\
, 'PctMathBenchmark11': 'percent_that_passedorabove_math_standard_11thgrade','TotNumBothBenchmark12':'number_that_passed_both_12thgrade' , 'PctBothBenchmark12': 'percent_that_passed_both_12thgrade','TotNumBothBenchmark11':'number_that_passed_both_11thgrade', 'PctBothBenchmark11':'percent_that_passed_both_11thgrade', 'Year':'year','Unnamed: 25':'unnamed_25'}

dfs = pd.read_csv(ca_sat_2019)

dfs.rename(columns = srename_map, inplace = True)
dfs.head()

county_district_school_code  county_code  district_code  school_code  \
0                 6.615981e+12          6.0       661598.0     630046.0   
1                 6.616061e+12          6.0       661606.0     634758.0   
2                 1.964733e+13         19.0      1964733.0    1930924.0   
3                 1.964733e+13         19.0      1964733.0    1931476.0   
4                 1.964733e+13         19.0      1964733.0    1931856.0   

  record_type              school_name        district_name  county_name  \
0           S  Colusa Alternative Home       Colusa Unified       Colusa   
1           S          Maxwell Sr High      Maxwell Unified       Colusa   
2           S      Belmont Senior High  Los Angeles Unified  Los Angeles   
3           S  Canoga Park Senior High  Los Angeles Unified  Los Angeles   
4           S     Whitman Continuation  Los Angeles Unified  Los Angeles   

   _12th_grade_enrollment  test_taker_count_12thgrade  ...  \
0                    18.0                         0.0  ...   
1                    29.0                        10.0  ...   
2                   206.0                       102.0  ...   
3                   227.0                       113.0  ...   
4                    18.0                        14.0  ...   

  number_that_passed_reading_writing_standard_11thgrade  \
0                                                NaN      
1                                                  *      
2                                                 42      
3                                                 97      
4                                                  *      

  percent_that_passed_reading_writing_standard_11thgrade  \
0                                                NaN       
1                                                  *       
2                                              24.14       
3                                              35.27       
4                                                  *       

  number_that_passedorabove_math_standard_11thgrade  \
0                                               NaN   
1                                                 *   
2                                                12   
3                                                37   
4                                                 *   

  percent_that_passedorabove_math_standard_11thgrade  \
0                                                NaN   
1                                                  *   
2                                               6.90   
3                                              13.45   
4                                                  *   

   number_that_passed_both_12thgrade  percent_that_passed_both_12thgrade  \
0                                NaN                                 NaN   
1                                  *                                   *   
2                                 14                               13.73   
3                                 18                               15.93   
4                                  *                                   *   

  number_that_passed_both_11thgrade percent_that_passed_both_11thgrade  \
0                               NaN                                NaN   
1                                 *                                  *   
2                                11                               6.32   
3                                35                              12.73   
4                                 *                                  *   

      year unnamed_25  
0  2018-19        NaN  
1  2018-19        NaN  
2  2018-19        NaN  
3  2018-19        NaN  
4  2018-19        NaN  

[5 rows x 26 columns]

In [17]:
# Change the column names using the `.rename()` function for ACT.

arename_map = {'CDS': 'county_district_school_code','CCode': 'county_code', 'CDCode': 'district_code','SCode':'school_code','RType': 'record_type', 'SName': 'school_name','DName': 'district_name', 'CName': 'county_name','Enroll12': '_12th_grade_enrollment', 'NumTstTakr': 'test_taker_count','AvgScrRead':'average_reading_score', 'AvgScrEng':'average_english_score', 'AvgScrMath':'average_math_score', 'AvgScrSci':'average_science_score', 'NumGE21':'number_of_composite_results_21andup', 'PctGE21':'percent_of_composite_results_21andup', 'Year':'year','Unnamed: 17':'unnamed_17'}
# Original column: renamed column

dfa = pd.read_csv(ca_act_2019)

dfa.rename(columns = arename_map, inplace = True)
dfa.head()

county_district_school_code  county_code  district_code  school_code  \
0                 3.366993e+13         33.0      3366993.0     129882.0   
1                 1.964212e+13         19.0      1964212.0    1995596.0   
2                 1.563776e+13         15.0      1563776.0    1530377.0   
3                 4.369666e+13         43.0      4369666.0    4333795.0   
4                 1.964733e+13         19.0      1964733.0    1935121.0   

  record_type                      school_name          district_name  \
0           S  21st Century Learning Institute       Beaumont Unified   
1           S      ABC Secondary (Alternative)            ABC Unified   
2           S      Abraham Lincoln Alternative  Southern Kern Unified   
3           S             Abraham Lincoln High       San Jose Unified   
4           S      Abraham Lincoln Senior High    Los Angeles Unified   

   county_name  _12th_grade_enrollment  test_taker_count  \
0    Riverside                    18.0               0.0   
1  Los Angeles                    58.0               0.0   
2         Kern                    18.0               0.0   
3  Santa Clara                   463.0              53.0   
4  Los Angeles                   226.0              19.0   

  average_reading_score average_english_score average_math_score  \
0                   NaN                   NaN                NaN   
1                   NaN                   NaN                NaN   
2                   NaN                   NaN                NaN   
3                    23                    22                 22   
4                    21                    20                 23   

  average_science_score number_of_composite_results_21andup  \
0                   NaN                                 NaN   
1                   NaN                                 NaN   
2                   NaN                                 NaN   
3                    23                                  34   
4                    22                                  11   

  percent_of_composite_results_21andup     year  unnamed_17  
0                                  NaN  2018-19         NaN  
1                                  NaN  2018-19         NaN  
2                                  NaN  2018-19         NaN  
3                                64.15  2018-19         NaN  
4                                57.89  2018-19         NaN

In [30]:
#dfs.county_district_school_code.unique()
#coded numbers-not needed-float

#dfs.county_code.unique()
#there is a NaN and other numbers appear to correpsond to the code above. There appear to be county listed\
#records as well as by school to show multple aggregates--float

#dfs.district_code.unique()
#there is a Nan and there appear to be unique numbers otherwise- in float

#dfs.school_code.unique()
#There is a Nan and there appear to be unique numbers otherwise-float

#dfs.record_type.unique()
# Four relevant values-stay as object but NAN

#dfs.school_name.unique()
#seems good except for one nan

#dfs.district_name.unique()
#stay as object and unique values

#dfs.county_name.unique()
#unique values-one nan and stay as object

#dfs._12th_grade_enrollment.unique()
#unique values and a nan

#dfs.test_taker_count_12thgrade.unique()
#unique values and a nan

#dfs.number_that_passed_reading_writing_standard_12thgrade.unique()
#there is a nan and * that need to be removed--*represents no reporting

#dfs.percent_that_passed_reading_writing_standard_12thgrade.unique()
#there is a nan and * that need to be removed--*represents no reporting

#dfs.number_that_passedorabove_math_standard_12thgrade.unique()
#unique values

#dfs.percent_that_passedorabove_math_standard_12thgrade.unique()
#nan and * need to be removed

#dfs.number_that_passed_both_12thgrade.unique()
#nan and * need to be removed

#dfs.percent_that_passed_both_12thgrade.unique()
#nan and * need to be removed.

#dfs._11th_grade_enrollment.unique()
#nan needs to be removed

#dfs.test_taker_count_11thgrade.unique()
#one nan needs to be removed

#dfs.number_that_reading_writing_standard_11thgrade.unique()
#Nan and * need to be removed

#dfs.percent_that_passed_reading_writing_standard_11thgrade.unique()
#nan and * need to be removed

#dfs.number_that_passedorabove_math_standard_11thgrade.unique()
#nan and * need to be removed

#dfs.percent_that_passedorabove_math_standard_11thgrade.unique()
#nan and * need to be removed

#dfs.number_that_passed_both_11thgrade.unique()
#nan and * need to be removed

#dfs.percent_that_passed_both_11thgrade.unique()
#nan and * need to be removed

#dfs.year.unique()
#nan and 2018-2019--useless column-remove

#dfs.unnamed_25.unique()
#Only nan-remove column

array([nan, '12', '37', '47', '176', '10', '84', '36', '1', '19', '91',
       '134', '48', '117', '16', '32', '259', '77', '264', '70', '113',
       '25', '193', '144', '41', '5', '546', '160', '9', '3', '14', '181',
       '15', '59', '356', '0', '97', '268', '8', '31', '44', '87', '18',
       '128', '57', '75', '180', '62', '29', '35', '94', '38', '42', '73',
       '210', '17', '191', '7', '145', '53', '28', '120', '22', '66',
       '336', '141', '63', '121', '140', '171', '262', '378', '150', '24',
       '292', '99', '106', '111', '115', '4', '86', '230', '256', '108',
       '135', '286', '21', '30', '11', '26', '43', '6', '146', '50',
       '151', '333', '196', '13', '218', '2', '54', '517', '89', '282',
       '56', '173', '285', '159', '46', '178', '60', '214', '81', '34',
       '215', '252', '313', '203', '200', '174', '100', '74', '168',
       '102', '61', '184', '343', '39', '110', '64', '20', '142', '280',
       '153', '393', '68', '183', '192', '51', '125', '279',

In [11]:
#dfa.county_district_school_code.unique()
#Nan needs to be removed

#dfa.county_code.unique()
#Nan needs to be removed

#dfa.district_code.unique()
#there is a nan and a 0, which doesnt make sense for a code identifier. Shouldnt effect results

#dfa.school_code.unique()
#unique values

#dfa.record_type.unique()
#nan needs to be removed

#dfa.school_name.unique()
#unique values.there is a state of California listing, which appears to be an aggregate.

#dfa.district_name.unique()
#there is a state of California listing, which appears to be an aggregate.

#dfa.county_name.unique()
#there is a state of California listing, which appears to be an aggregate. Also a nan

#dfa._12th_grade_enrollment.unique()
#there is a nan that needs to be removed

#dfa.test_taker_count.unique()
#there appears to be a nan

#dfa.average_reading_score.unique()
#there appears to be a nan and * that needs to be removed

#dfa.average_english_score.unique()
#there appears to be a nan and * that needs to be removed.

#dfa.average_math_score.unique()
#there appears to be a nan and * that needs to be removed

#dfa.average_science_score.unique()
#there appears to be a nan and * that needs to be removed

#dfa.number_of_composite_results_21andup.unique()
#there appears to be a nan and * that needs to be removed

#dfa.percent_of_composite_results_21andup.unique()
#there appears to be a nan and * that needs to be removed

#dfa.year.unique()
#all 2018-2019-not needed also a nan

#dfa.unnamed_17.unique()
#all nans - remove column

In [18]:
# Deleting unnecessary columns including unnamed as they are all nan's and the year is the same for all datafrom both as thas theand year
dfs.drop(columns=['unnamed_25', 'year'],inplace=True)
dfa.drop(columns=['unnamed_17', 'year'], inplace=True)


In [133]:
#asterisk_release_mask = dfs['number_that_passed_reading_writing_standard_12thgrade'].str.contains(*)
#dfs.loc[asterisk_release_mask, 'number_that_passed_reading_writing_standard_12thgrade'] = np.NaN
#dfs.head(15)


SyntaxError: invalid syntax (<ipython-input-133-774c5fde4bf8>, line 1)

In [19]:
#turning asterisks into NaN's
dfs = dfs.replace({'*' : np.NaN})
dfa = dfa.replace({'*': np.NaN})
dfs.head()

county_district_school_code  county_code  district_code  school_code  \
0                 6.615981e+12          6.0       661598.0     630046.0   
1                 6.616061e+12          6.0       661606.0     634758.0   
2                 1.964733e+13         19.0      1964733.0    1930924.0   
3                 1.964733e+13         19.0      1964733.0    1931476.0   
4                 1.964733e+13         19.0      1964733.0    1931856.0   

  record_type              school_name        district_name  county_name  \
0           S  Colusa Alternative Home       Colusa Unified       Colusa   
1           S          Maxwell Sr High      Maxwell Unified       Colusa   
2           S      Belmont Senior High  Los Angeles Unified  Los Angeles   
3           S  Canoga Park Senior High  Los Angeles Unified  Los Angeles   
4           S     Whitman Continuation  Los Angeles Unified  Los Angeles   

   _12th_grade_enrollment  test_taker_count_12thgrade  ...  \
0                    18.0                         0.0  ...   
1                    29.0                        10.0  ...   
2                   206.0                       102.0  ...   
3                   227.0                       113.0  ...   
4                    18.0                        14.0  ...   

  _11th_grade_enrollment test_taker_count_11thgrade  \
0                   18.0                        0.0   
1                   26.0                        6.0   
2                  219.0                      174.0   
3                  333.0                      275.0   
4                   17.0                        5.0   

  number_that_passed_reading_writing_standard_11thgrade  \
0                                                NaN      
1                                                NaN      
2                                                 42      
3                                                 97      
4                                                NaN      

  percent_that_passed_reading_writing_standard_11thgrade  \
0                                                NaN       
1                                                NaN       
2                                              24.14       
3                                              35.27       
4                                                NaN       

   number_that_passedorabove_math_standard_11thgrade  \
0                                                NaN   
1                                                NaN   
2                                                 12   
3                                                 37   
4                                                NaN   

   percent_that_passedorabove_math_standard_11thgrade  \
0                                                NaN    
1                                                NaN    
2                                               6.90    
3                                              13.45    
4                                                NaN    

  number_that_passed_both_12thgrade percent_that_passed_both_12thgrade  \
0                               NaN                                NaN   
1                               NaN                                NaN   
2                                14                              13.73   
3                                18                              15.93   
4                               NaN                                NaN   

  number_that_passed_both_11thgrade percent_that_passed_both_11thgrade  
0                               NaN                                NaN  
1                               NaN                                NaN  
2                                11                               6.32  
3                                35                              12.73  
4                               NaN                                NaN  

[5 rows x 24 columns]

In [20]:
# Dropping rows with NaN's in these particular columns
dfa1 = dfa.dropna(subset=['_12th_grade_enrollment', 'test_taker_count', 'average_reading_score'\
, 'average_english_score', 'average_math_score', 'average_science_score', 'number_of_composite_results_21andup'\
, 'percent_of_composite_results_21andup']).copy()

dfs1 = dfs.dropna(subset=['_12th_grade_enrollment','test_taker_count_12thgrade','number_that_passed_reading_writing_standard_12thgrade'\
,'percent_that_passed_reading_writing_standard_12thgrade','number_that_passedorabove_math_standard_12thgrade'\
,'percent_that_passedorabove_math_standard_12thgrade','number_that_passed_both_12thgrade'\
,'percent_that_passed_both_12thgrade','_11th_grade_enrollment','test_taker_count_11thgrade'\
,'number_that_passed_reading_writing_standard_11thgrade','percent_that_passed_reading_writing_standard_11thgrade'\
,'number_that_passedorabove_math_standard_11thgrade','percent_that_passedorabove_math_standard_11thgrade','number_that_passed_both_11thgrade'\
,'percent_that_passed_both_11thgrade']).copy()

#Source:classmate and https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

In [44]:
print (dfs1.isnull())

      county_district_school_code  county_code  district_code  school_code  \
2                           False        False          False        False   
3                           False        False          False        False   
5                           False        False          False        False   
6                           False        False          False        False   
9                           False        False          False        False   
...                           ...          ...            ...          ...   
2574                        False        False          False        False   
2575                        False        False          False        False   
2576                        False        False          False        False   
2577                        False        False          False        False   
2578                        False        False          False        False   

      record_type  school_name  district_name  county_name  \
2

In [45]:
print (dfa1.isnull())

      county_district_school_code  county_code  district_code  school_code  \
3                           False        False          False        False   
4                           False        False          False        False   
5                           False        False          False        False   
6                           False        False          False        False   
13                          False        False          False        False   
...                           ...          ...            ...          ...   
2298                        False        False          False        False   
2299                        False        False          False        False   
2302                        False        False          False        False   
2303                        False        False          False        False   
2306                        False        False          False        False   

      record_type  school_name  district_name  county_name  \
3

In [23]:
dfs1['county_name'].sort_values()

1666    Alameda
1085    Alameda
1671    Alameda
1248    Alameda
553     Alameda
         ...   
2022       Yuba
2233       Yuba
1359       Yuba
765        Yuba
1721       Yuba
Name: county_name, Length: 1639, dtype: object

In [24]:
dfa1['county_name'].sort_values()
#print(ca_sat_2019.isnull().sum())

1711    Alameda
643     Alameda
1131    Alameda
461     Alameda
1809    Alameda
         ...   
2268       Yuba
958        Yuba
892        Yuba
1642       Yuba
1643       Yuba
Name: county_name, Length: 1421, dtype: object

In [50]:
dfs1._12th_grade_enrollment = dfs1._12th_grade_enrollment.astype(np.float64)

dfs1.test_taker_count_12thgrade = dfs1.test_taker_count_12thgrade.astype(np.float64)

dfs1.number_that_passed_reading_writing_standard_12thgrade = dfs1.number_that_passed_reading_writing_standard_12thgrade.astype(np.float64)

dfs1.number_that_passed_reading_writing_standard_12thgrade = dfs1.number_that_passed_reading_writing_standard_12thgrade.astype(np.float64)

dfs1.number_that_passed_reading_writing_standard_12thgrade = dfs1.number_that_passed_reading_writing_standard_12thgrade.astype(np.float64)

dfs1.percent_that_passed_reading_writing_standard_12thgrade = dfs1.percent_that_passed_reading_writing_standard_12thgrade.astype(np.float64)

dfs1.number_that_passedorabove_math_standard_12thgrade = dfs1.number_that_passedorabove_math_standard_12thgrade.astype(np.float64)

dfs1.percent_that_passedorabove_math_standard_12thgrade = dfs1.percent_that_passedorabove_math_standard_12thgrade.astype(np.float64)

dfs1.number_that_passed_both_12thgrade = dfs1.number_that_passed_both_12thgrade.astype(np.float64)

dfs1.percent_that_passed_both_12thgrade = dfs1.percent_that_passed_both_12thgrade.astype(np.float64)

dfs1._11th_grade_enrollment = dfs1._11th_grade_enrollment.astype(np.float64)

dfs1.test_taker_count_11thgrade = dfs1.test_taker_count_11thgrade.astype(np.float64)

dfs1.number_that_passed_reading_writing_standard_11thgrade = dfs1.number_that_passed_reading_writing_standard_11thgrade.astype(np.float64)

dfs1.percent_that_passed_reading_writing_standard_11thgrade = dfs1.percent_that_passed_reading_writing_standard_11thgrade.astype(np.float64)

dfs1.number_that_passedorabove_math_standard_11thgrade = dfs1.number_that_passedorabove_math_standard_11thgrade.astype(np.float64)

dfs1.percent_that_passedorabove_math_standard_11thgrade = dfs1.number_that_passedorabove_math_standard_11thgrade.astype(np.float64)

dfs1.number_that_passed_both_11thgrade = dfs1.number_that_passed_both_11thgrade.astype(np.float64)

dfs1.percent_that_passed_both_11thgrade = dfs1.percent_that_passed_both_11thgrade.astype(np.float64)

dfs1.dtypes  
#percent_that_passedorabove_math_standard_11thgrade','TotNumBothBenchmark12':'number_that_passed_both_12thgrade' , 'PctBothBenchmark12': 'percent_that_passed_both_12thgrade','

TypeError: float() argument must be a string or a number, not 'generator'

In [23]:

dfa1._12th_grade_enrollment = dfa1._12th_grade_enrollment.astype(np.float64)

dfa1.test_taker_count = dfa1.test_taker_count.astype(np.float64)

dfa1.average_reading_score = dfa1.average_reading_score.astype(np.float64)

dfa1.average_english_score = dfa1.average_english_score.astype(np.float64)

dfa1.average_math_score = dfa1.average_math_score.astype(np.float64)

dfa1.average_science_score = dfa1.average_science_score.astype(np.float64)

dfa1.average_science_score = dfa1.average_science_score.astype(np.float64)

dfa1.percent_of_composite_results_21andup = dfa1.percent_of_composite_results_21andup.astype(np.float64)

dfa1.number_of_composite_results_21andup = dfa1.number_of_composite_results_21andup.astype(np.float64)
#','').\
#astype


#', 'a',', '', 'number_of_composite_results_21andup
#https://stackoverflow.com/questions/17950374/converting-a-column-within-pandas-dataframe-from-int-to-string
#print (ca_act_2019.isnull().sum())
dfa1.dtypes

county_district_school_code             float64
county_code                             float64
district_code                           float64
school_code                             float64
record_type                              object
school_name                              object
district_name                            object
county_name                              object
_12th_grade_enrollment                  float64
test_taker_count                        float64
average_reading_score                   float64
average_english_score                   float64
average_math_score                      float64
average_science_score                   float64
number_of_composite_results_21andup     float64
percent_of_composite_results_21andup    float64
dtype: object

In [48]:
# Changing Percents to Decimals using function above
dfa1.percent_of_composite_results_21andup = (percent_decimal(x) for x in dfa1.percent_of_composite_results_21andup)

dfs1.percent_that_passed_reading_writing_standard_12thgrade=(percent_decimal(x) for x in dfs1.percent_that_passed_reading_writing_standard_12thgrade)

#dfs1.percent_that_passedorabove_math_standard_12thgrade=[percent_decimal(x) for x in dfs1.percent_that_passedorabove_math_standard_12thgrade]

dfs1.percent_that_passed_reading_writing_standard_11thgrade=(percent_decimal(x) for x in dfs1.percent_that_passed_reading_writing_standard_11thgrade)

dfs1.percent_that_passedorabove_math_standard_11thgrade = (percent_decimal(x) for x in dfs1.percent_that_passedorabove_math_standard_11thgrade)

dfs1.percent_that_passed_both_12thgrade= (percent_decimal(x) for x in dfs1.percent_that_passed_both_12thgrade)

dfs1.percent_that_passed_both_11thgrade= (percent_decimal(x) for x in dfs1.percent_that_passed_both_11thgrade)

dfa1.percent_of_composite_results_21andup

dfa1.dtypes


county_district_school_code             float64
county_code                             float64
district_code                           float64
school_code                             float64
record_type                              object
school_name                              object
district_name                            object
county_name                              object
_12th_grade_enrollment                  float64
test_taker_count                        float64
average_reading_score                    object
average_english_score                    object
average_math_score                       object
average_science_score                    object
number_of_composite_results_21andup      object
percent_of_composite_results_21andup     object
dtype: object

M



#All the asteriks, which indicate that the scores were not reported due to privacy\
#concerns as the sample was so small at that school that the children would be identified,\
#have been converted to NaN's.
ca_act_2019_clean1= ca_act_2019.replace({'*' : np.NaN})
ca_sat_2019_clean1= ca_sat_2019.replace({'*': np.NaN})

In [109]:
ca_sat_2019_clean1['NumERWBenchmark12'].sort_values()

707       1
1863      1
1841     10
735      10
816      10
       ... 
1968    NaN
1970    NaN
1972    NaN
1977    NaN
1978    NaN
Name: NumERWBenchmark12, Length: 1602, dtype: object

In [11]:
# Given that my datasets are greater than 2000 each, I have filtered out the NaN's
#from my datasets. I learned of this function from a classmate.
ca_act_2019_clean2 = ca_act_2019_clean1.dropna().copy()
ca_sat_2019_clean2 = ca_sat_2019_clean1.dropna().copy()

In [113]:
#ca_act_2019_clean.isnull().sum()
ca_act_2019_clean2['PctGE21'].sort_values()

820      0.00
2146     0.00
88       0.00
539      0.00
129      0.00
        ...  
1006    96.18
1992    96.69
1025    97.17
1809    97.30
2271    98.18
Name: PctGE21, Length: 1017, dtype: object

In [115]:
#ca_sat_2019_clean.isnull().sum()
ca_sat_2019_clean2['PctBothBenchmark11'].sort_values()

454      0.00
940      0.00
60       0.00
1641     0.00
1078     0.00
        ...  
401     98.01
541     98.95
1747    99.19
144     99.47
1100    99.68
Name: PctBothBenchmark11, Length: 1199, dtype: object

In [118]:
#ca_sat_2019_clean2.head()
#ca_sat_2019_clean2[NumERWBenchmark12]=ca_sat_2019_clean2( iny)
ca_sat_2019_clean1.NumERcolumn_name = df.column_name.astype(np.int64)
#percent_decimal('5%')
https://stackoverflow.com/questions/17950374/converting-a-column-within-pandas-dataframe-from-int-to-string

SName                      DName        CName  \
2          Belmont Senior High        Los Angeles Unified  Los Angeles   
3      Canoga Park Senior High        Los Angeles Unified  Los Angeles   
5       Foshay Learning Center        Los Angeles Unified  Los Angeles   
6           Mission Viejo High  Saddleback Valley Unified       Orange   
9  New Open World Academy K-12        Los Angeles Unified  Los Angeles   

   Enroll12  NumTSTTakr12 NumERWBenchmark12 PctERWBenchmark12  \
2     206.0         102.0                31             30.39   
3     227.0         113.0                54             47.79   
5     166.0         106.0                68             64.15   
6     562.0         190.0               161             84.74   
9      85.0          68.0                17             25.00   

  NumMathBenchmark12 PctMathBenchmark12  Enroll11  NumTSTTakr11  \
2                 14              13.73     219.0         174.0   
3                 18              15.93     333.0         275.0   
5                 36              33.96     187.0         183.0   
6                138              72.63     503.0         223.0   
9                  6               8.82      79.0          79.0   

  NumERWBenchmark11 PctERWBenchmark11 NumMathBenchmark11 PctMathBenchmark11  \
2                42             24.14                 12               6.90   
3                97             35.27                 37              13.45   
5                91             49.73                 47              25.68   
6               206             92.38                176              78.92   
9                22             27.85                 10              12.66   

  TotNumBothBenchmark12 PctBothBenchmark12 TotNumBothBenchmark11  \
2                    14              13.73                    11   
3                    18              15.93                    35   
5                    36              33.96                    44   
6                   132              69.47                   172   
9                     6               8.82                     6   

  PctBothBenchmark11  
2               6.32  
3              12.73  
5              24.04  
6              77.13  
9               7.59

### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|ACT/SAT|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in the dataframe.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value 
        - *Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest participation rates for the 2017, 2019, or 2019 SAT and ACT?
        - Which states have the highest and lowest mean total/composite scores for the 2017, 2019, or 2019 SAT and ACT?
        - Do any states with 100% participation on a given test have a rate change year-to-year?
        - Do any states show have >50% participation on *both* tests each year?
        - Which colleges have the highest median SAT and ACT scores for admittance?
        - Which California school districts have the highest and lowest mean test scores?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
#Code:

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [ ]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

Don't forget to create your README!

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!